### PR TITLE
Added possibility to use a page size preset as argument for b.size()

### PR DESCRIPTION
**Pull request by trych**
Hey there,
just had the idea that it would be nice if instead of using values to set the page size with `b.size()`, you could also set the size with the name of a page size preset like `"A4"` or `"Letter"`, etc.

Was easy to implement, so I decided to create a pull request right away. Hope the idea makes sense to you guys as well.

However, before anybody goes ahead to merge this, one question: Should we also allow yet another parameter to set the orientation? With the implementation I just did, it simply keeps the orientation. That means if I have a landscape `A4` document and set it to `iPhone` format via `b.size("iPhone")` I will get a landscape iPhone sized document automatically.

Should we allow an extra parameter like `b.PORTRAIT` and `b.LANDSCAPE` (in which case I would create a new pull request with that implemented)?

So usage would look like this:

`b.size("A3", b.LANDSCAPE);`

### CODE REVIEW

**Comment by b-g** [comment]
Great! 👍

Yes to `b.PORTRAIT`, `b.LANDSCAPE` and `b.size("A3", b.LANDSCAPE)`; from my end.

File: basil.js [hunk 1]
```diff
@@ -1494,44 +1494,56 @@ pub.doc = function(doc) {
 /**
  * Sets the size of the current document, if arguments are given.
  * If only one argument is given, both the width and the height are set to this value.
+ * Alternatively a string can be given as the only argument to apply an existing page size preset ("A4", "Letter" etc.).
```
Comment (fabianmoronzirfas) [inline basil.js]: - What are the possible presets?
- Can we list them in the docs?
- What about localization?

File: basil.js [hunk 2]
```diff
@@ -1494,44 +1494,56 @@ pub.doc = function(doc) {
...
 
-  if(arguments.length === 1) {
+  var doc = currentDoc();
+
+  if(arguments.length === 1 && typeof widthOrPageSize === "string") {
+
+    try {
+      doc.documentPreferences.pageSize = widthOrPageSize;
+    } catch (e) {
+      b.error("b.size(), could not find a page size preset named \"" + widthOrPageSize + "\".");
```
Comment (fabianmoronzirfas) [inline basil.js]: Is there a way to get all the existing preset names and list them here?

Comment (fabianmoronzirfas) [inline basil.js]: When running this code:  

```js
var presets = app.documentPresets;
for (var i = presets.length - 1; i >= 0; i--) {
  $.writeln(presets[i].name);
}
```

I only get:  

    [Standard]

as output. Am I missing someting?

**Comment by fabianmoronzirfas** [comment]
The result of these `app.documentPresets` is what is defined here  

bildschirmfoto 2017-07-14 um 08 38 57

File: basil.js (hunk 1 above)
Comment (trych) [inline basil.js]: (Sorry, for the late response, I am on holiday now 😎 )

No, we cannot list them in the docs, as every user can define his own presets with own names (and probably there are a lot of different presets over different versions). So, I would just leave them out and I just put "A4" and "Letter" in there as examples, so people understand right away, what we are talking about.

File: basil.js (hunk 2 above)
Comment (trych) [inline basil.js]: Yes, that's not the page size presets, that's the document presets that also saves stuff like columns and gutter width etc. Since those are not related to page size, I would just leave them out and focus on page size only. Since I don't know how to retrieve all those, I would just leave them out of the error message.

File: basil.js (hunk 1 above)
Comment (fabianmoronzirfas) [inline basil.js]: But how can I list those page size presets? Can you point me to some code?

Comment (trych) [inline basil.js]: I don't think you can. I have looked around for a bit as well, and couldn't find anything. You can just catch an error to find out that a page size preset does not exist, afaik.

Comment (fabianmoronzirfas) [inline basil.js]: Okay. At least I found where the ones the user creates are stored.  

    "/Users/$USERNAME/Library/Preferences/Adobe InDesign/Version 12.0/de_DE/Page Sizes"

It would be cool to parse these xml files and give this user a hint what presets he made by himself. I'll investigate a bit further to find them. You can create them by hitting `⌘ ⇧ p`  

bildschirmfoto 2017-07-18 um 07 35 31

Comment (fabianmoronzirfas) [inline basil.js]: sorry it is `⌥ ⌘ p`

Comment (fabianmoronzirfas) [inline basil.js]: I got some update from the Adobe Dev Slack channel by Dirk Becker. These values are not scriptable. 😢  but made a dump of the relevant parts for us. So we have at least a list of them.
[InDesign_kService_PageSizes.c.txt](https://github.com/basiljs/basil.js/files/1154980/InDesign_kService_PageSizes.c.txt)

**Comment by trych** [comment]
Ok, I implemented the page orientation feature, added tests and added quick examples to the docs.

@fabianmoronzirfas I would leave listing of existing page presets out. I don't think it's very important at all, but on the other hand it would be very complicated to include. Also, there are different presets from version to version. The ones that you linked in the text file are not matching the ones that I have in my CS6 installation at all, for example.

This passes all tests and is ready for review. 🔍 🐞

**Comment by fabianmoronzirfas** [comment]
I like what you implemented but find it a bit obscure that we only can't verify if it works by catching an error.

**Comment by trych** [comment]
I agree that it would be nicer to do it another way, but as long as it is not possible …
Or do you know another way?

I mean the user will not know, so it should not matter for user experience. There are a few `try … catch` statements in basil, so I think it should be fine.

**Comment by fabianmoronzirfas** [comment]
I don't know another way. In the outcome it is irrelevant for the user. try catch just feels like using a crowbar to open a door. 😁

**Comment by ffd8** [comment]
Sounds great, but to be honest, I (+ students) never change the document size in code. It's decided before starting, however when doing such a change, this would be very useful. I also like the idea of `b.PORTRAIT` and `b.LANDSCAPE` as optional inputs (perhaps defaulting with portrait?).

PS - if you don't already have a list of page sizes, perhaps this one can help:
`b.inspect(app.printerPresets[0].paperSizeList);`?

**Comment by trych** [comment]
> Sounds great, but to be honest, I (+ students) never change the document size in code.

That's true, but I think the `b.size()` function is actually really useful, if you want to share your basil scripts and have a certain target size in mind. The person who downloads your script would not know how to set up the document, so it is useful to have the script itself take care of setting the size.

I think the printerPreset's paper size list only lists paper formats specific to a printer and does not reflect the page size presets that are available under the document preferences.

Btw. default is neither `b.PORTRAIT` nor `b.LANDSCAPE` but rather the orientation that the document already has. So if you use `b.size("A3")` on a document that is currently in portrait A4, it will stay portrait.

**Comment by ffd8** [comment]
Aha – that's a good example for using `b.size()`. 
And true, that's the paper sizes for printing.. just thought it might be a useful start.

**Comment by trych** [comment]
@b-g @fabianmoronzirfas @ffd8 

This has been waiting here for a while, but should actually be good to go as well, unless you guys have some objections?
